docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,205 @@
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
+	/api/v1/nodeMetrics/myNode/[derived|raw] 
+
+The derived metrics include the mean, max and a few percentiles of the list of
+values.
+
+The raw metrics are the values themselves, thus the data will be much heavier.
+
+In the future, if any uses cases are found that would benefit from RC,
+namespace or service aggregation, metrics at those levels could also be
+exposed taking advantage of the fact that Heapster already does aggregation
+and metrics for them.
+
+## Data gathering
+
+We can either use a Push or a Poll based system:
+
+* Push - each node would periodically POST their derived and raw metrics to
+  the API server. Kubectl and the scheduler would watch the API server to get
+  the latest metrics.
+
+* Poll - heapster would periodically poll a kubelet API endpoint on each node
+  to fetch its node and pods metrics. Then Heapster will have to implement the
+  APIs mentioned above.
+
+Our initial plan is to follow the push based system.
+
+The API server will not hold any of this data directly. For our initial
+purposes, it will hold the most recent metrics obtained from each node in
+etcd. Then, when polled for metrics, the API server would only serve said most
+recent data per node.
+
+If a pod gets deleted, the API server will get rid of any metrics it may
+currently be holding for it.
+
+The clients watching the metrics data may cache it for longer periods of time.
+The clearest example would be Heapster.
+
+In the future, we might want to store the metrics differently:
+
+* via heapster - Since heapster keeps data for a period of time, we could
+  redirect requests to the API server to heapster instead of using etcd. This
+  would also allow serving metrics other than the latest ones.
+
+An edge case that this proposal doesn't take into account is kubelets being
+restarted. If any of them are, with a simple implementation they would lose
+historical data and thus take hours to gather enough information to provide
+relevant metrics again. We might want to use persistent storage directly or in
+the future to improve that situation.
+
+## Data structure
+
+```Go
+type DerivedPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	PodMetrics ResourcesMetrics
+	// the key is the container name
+	Containers []struct {
+		ContainerReference *Container
+		Metrics            ResourcesMetrics
+	}
+}
+
+type DerivedNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      ResourcesMetrics
+	SystemContainers []struct {
+		ContainerReference *Container
+		Metrics            ResourcesMetrics
+	}
+}
+
+type ResourcesMetrics map[resource.Type]DerivedMetrics
```

Comment:
Do we need watch?  Is someone watching the metrics to see them change?
Since the pushing is coming on a fixed interval that means that the watch
would also get fixed intervals, even if nothing meaningful changes.

On Wed, Jul 29, 2015 at 6:01 PM, Daniel Martí <notifications@github.com>
wrote:

> In docs/proposals/compute-resource-metrics-api.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35817295>
> :
>
> > +		ContainerReference *Container
> > +		Metrics            ResourcesMetrics
> > +	}
> > +}
> > +
> > +type DerivedNodeMetrics struct {
> > +	TypeMeta
> > +	ObjectMeta // should have node name
> > +	NodeMetrics      ResourcesMetrics
> > +	SystemContainers []struct {
> > +		ContainerReference *Container
> > +		Metrics            ResourcesMetrics
> > +	}
> > +}
> > +
> > +type ResourcesMetrics map[resource.Type]DerivedMetrics
>
> @bgrant0607 <https://github.com/bgrant0607> should this
> map[resource.Type]... and the map[time.Duration]... below be converted to
> lists too?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35817295>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift